Shape_detection/include/CGAL/Shape_detection/Region_growing/Triangle_mesh/One_ring_neighbor_query.h

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2018 INRIA Sophia-Antipolis (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+//
+// Author(s)     : Florent Lafarge, Simon Giraudot, Thien Hoang, Dmitry Anisimov
+//
+
+#ifndef CGAL_SHAPE_DETECTION_REGION_GROWING_TRIANGLE_MESH_ONE_RING_NEIGHBOR_QUERY_H
+#define CGAL_SHAPE_DETECTION_REGION_GROWING_TRIANGLE_MESH_ONE_RING_NEIGHBOR_QUERY_H
+
+#include <CGAL/license/Shape_detection.h>
+
+// Internal includes.
+#include <CGAL/Shape_detection/Region_growing/internal/property_map.h>
+
+namespace CGAL {
+namespace Shape_detection {
+namespace Triangle_mesh {
+
+  /*!
+    \ingroup PkgShapeDetectionRGOnMesh
+
+    \brief Edge-adjacent faces connectivity in a triangle mesh.
+
+    This class returns all faces, which are edge-adjacent to a query face in a
+    triangle mesh being a `TriangleMesh`.
+
+    \tparam TriangleMesh
+    a model of `FaceListGraph`
+
+    \tparam FaceRange
+    a model of `ConstRange` whose iterator type is `RandomAccessIterator` and
+    value type is the face type of a triangle mesh
+
+    \cgalModels `NeighborQuery`
+  */
+  template<typename TriangleMesh
+#ifndef CGAL_NO_DEPRECATED_CODE
+  , typename FaceRange = void
+#endif
+  >
+  class One_ring_neighbor_query
+  {
+    using Face_to_index_map = typename boost::property_map<TriangleMesh, CGAL::dynamic_face_property_t<std::size_t> >::const_type;
+  public:
+    /// \cond SKIP_IN_MANUAL
+    using face_descriptor = typename boost::graph_traits<TriangleMesh>::face_descriptor;
+    using Face_graph = TriangleMesh;
+
+    using Item = face_descriptor;
+    using Region = std::vector<Item>;
+    /// \endcond
+
+    /// \name Initialization
+    /// @{
+
+    /*!
+      \brief initializes all internal data structures.
+
+      \param pmesh
+      an instance of a `TriangleMesh` that represents a triangle mesh
+
+      \pre `faces(pmesh).size() > 0`
+    */
+    One_ring_neighbor_query(
+      const TriangleMesh& tmesh) :
+    m_face_graph(tmesh),
+    m_face_to_index_map(get(CGAL::dynamic_face_property_t<std::size_t>(), tmesh))
+    {
+      m_face_range.reserve(num_faces(tmesh)); // a bit larger if has garbage
+      for (face_descriptor f : faces(tmesh))
+      {
+        put(m_face_to_index_map, f, m_face_range.size());
+        m_face_range.push_back(f);
+      }
+      CGAL_precondition(m_face_range.size() > 0);
+    }
```

Comment:
I have just realized that this constructor does not enable users to run region growing on a subset of faces in the mesh. I would add second constructor 
```
One_ring_neighbor_query(const TriangleMesh &tmesh, const FaceRange &faces) { }
```
where users can provide their own subset of faces. I think it would be simpler than first creating a filtered surface mesh and then run region growing on it. That will also keep right indices with respect to the whole range of faces.